src/Temporalio.Extensions.Hosting/ITemporalClientProvider.cs

```diff
@@ -0,0 +1,16 @@
+using System.Threading.Tasks;
+using Temporalio.Client;
+
+namespace Temporalio.Extensions.Hosting
+{
+    /// <summary>
+    /// Interface for providing a lazily instantiated <see cref="ITemporalClient" />.
+    /// </summary>
+    public interface ITemporalClientProvider
+    {
+        /// <summary>
+        /// Get the client.
+        /// </summary>
+        Task<ITemporalClient> GetClient();
```

Comment:
@mrtristan @cretz 
Is it possible to create a HostedService that will start a client and set it, for example, in `ITemporalClientProvider.Client`? I'm not sure about the practical necessity, just curious, and I don't like that `GetClient` returns `Task<T>` 😊, even though initialization happens only once. It seems to me that it would be better to use `ValueTask<T>`..

Sketch with HostedService:
```csharp
public interface ITemporalClientProvider
{
  public ITemporalClient Client { get; }
}

public TemporalClientProviderInitializer(
  TemporalClientProvider provider,
  ITemporalClientFactory factory,
  ... other dependencies)
{
  _provider = provider;
  _factory = factory;
  ...
}
public async Task StartAsync(...)
{
  var client = await _factory.Create();
  _provider.Client = client;
}

services.AddSingleton<TemporalClientProvider>();
services.AddSingleton<ITemporalClientProvider>(provider => provider.GetRequiredService<TemporalClientProvider>());
services.AddHostedService<TemporalClientProviderInitializer>();
services.AddHostedService<TemporalWorkerService>(); // must be after

services.AddTemporalClientFactory(...); // user defined ITemporalClientFactory
```
Or maybe using IApplicationBuilder would be more explicit when the client is created, but it has the downside of using Minimal API:
```csharp
await app.InitializeTemporalClient();

public static Task InitializeTemporalClient(this IApplicationBuilder app)
{
  var factory = app.Services.GetRequiredService<ITemporalClientFactory>();
  var provider = app.Services.GetRequiredService <TemporalClientProvider>();
  provider.Client = await factory.Create();
}
```

Reply:
@devbased 
chad pointed out to me what i was missing regarding the capabilities here to have a DI'd yet lazily instantiated client which was my fundamental need. if you can live with a few clients being created (between the workers and your own usages) it seems fine to just get fancy in injecting/configuring the `TemporalClientConnectOptions` (or at least that covered my need).

https://temporalio.slack.com/archives/C012SHMPDDZ/p1688156494059699

Reply:
@devbased - I am going to open up a PR hopefully later today on my thoughts here. It will probably include the ability to configure `TemporalClientConnectOptions` + `TemporalClientProvider`, and a `TemporalClient.ConnectLazy` which is non-async (and DI extensions for that too).

Reply:
Opened https://github.com/temporalio/sdk-dotnet/pull/98, take a look